lib/rules/destructuring-props-argument.js

```diff
@@ -0,0 +1,89 @@
+/**
+ * @fileoverview Rule to forbid or enforce destructuring props argument.
+ **/
+'use strict';
+
+const Components = require('../util/Components');
+
+const DEFAULT_OPTIONS = {
+  SFC: 'always'
+};
+
+module.exports = {
+  meta: {
+    docs: {
+      description: 'Forbid or enforce destructuring props argument',
+      category: 'Stylistic Issues',
+      recommended: false
+    },
+    schema: [{
+      definitions: {
+        value: {
+          enum: [
+            'always',
+            'never',
+            'ignore'
+          ]
+        }
+      },
+      type: 'object',
+      properties: {
+        SFC: {$ref: '#/definitions/value'}
+      },
+      additionalProperties: false
+    }]
+  },
+
+  create: Components.detect((context, components) => {
+    const configuration = context.options[0] || {};
+    const options = {SFC: configuration.SFC || DEFAULT_OPTIONS.SFC};
+
+
+    /**
+    * Checks if a prop is being assigned a value props.bar = 'bar'
+    * @param {ASTNode} node The AST node being checked.
+    * @returns {Boolean}
+    */
+
+    function isAssignmentToProp(node) {
+      return (
+        node.parent &&
+        node.parent.type === 'AssignmentExpression' &&
+        node.parent.left === node
+      );
+    }
+    /**
+     * @param {ASTNode} node We expect either an ArrowFunctionExpression,
+     *   FunctionDeclaration, or FunctionExpression
+     */
+    function handleStatelessComponent(node) {
+      const destructuring = node.params && node.params[0] && node.params[0].type === 'ObjectPattern';
+      if (destructuring && components.get(node) && options.SFC === 'never') {
+        context.report({
+          node: node,
+          message: 'Should never use destructuring props assignment in SFC argument'
+        });
+      }
+    }
+
+    return {
+
+      FunctionDeclaration: handleStatelessComponent,
+
+      ArrowFunctionExpression: handleStatelessComponent,
+
+      FunctionExpression: handleStatelessComponent,
+
+      MemberExpression: function(node) {
+        const relatedComponent = components.get(context.getScope(node).block);
+        const isStatelessFunctionUsage = node.object.name === 'props' && !isAssignmentToProp(node);
+        if (relatedComponent && isStatelessFunctionUsage && options.SFC === 'always') {
+          context.report({
+            node: node,
+            message: 'Should use destructuring props assignment in SFC argument'
```

Comment:
`s/Should/Must/g`?

Reply:
hm, what do you mean?

Reply:
it's not that you *should* use it, it's that you *must* because of the rule